### PR TITLE
add max tracked block config variable

### PR DESCRIPTION
**Pull request by danielradu10**
## Reasoning behind the pull request
- At the moment, there is no check  in case **too many blocks are tracked** in the selection tracker. In this PR, we manage this by adding a new config variable, an upper bound - **MaxTrackedBlocks** - for how many blocks could be in the tracker and we **treat the case when this dimension is reached**.
- Also, we check with some integration tests that **proposing some empty blocks** would not produce **any not-wanted behavior.**
  
## Proposed changes
- added a new **config** variable **MaxTrackedBlocks** and a variable on the **instance** of the **SelectionTracker**.
- **empty the map of tracked blocks** and log in case that upper bound is reached.
- added integration tests for this scenario.
- added integration tests for the scenarios when we propose empty blocks.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: txcache/errors.go [hunk 1]
```diff
@@ -19,3 +19,4 @@ var errNonceGap = errors.New("nonce gap")
 var errExceededBalance = errors.New("exceeded balance")
 var errNonceNotSet = errors.New("nonce not set")
 var errReceivedLastNonceNotSet = errors.New("received last nonce not set")
+var errInvalidMaxTrackedBlocks = errors.New("invalid max tracked blocks")
```
Comment (andreibancioiu) [inline txcache/errors.go]: `var ErrBadMaxTrackedBlocks = errors.New("bad max tracked blocks")` 

vs.

`var errInvalidMaxTrackedBlocks = errors.New("invalid max tracked blocks")`

Maybe use the same error message (nitpicking).

File: dataRetriever/txpool/argShardedTxPool.go [hunk 2]
```diff
@@ -53,5 +53,9 @@ func (args *ArgShardedTxPool) verify() error {
 		return fmt.Errorf("%w: MaxNumBytesPerSenderUpperBound is not valid", dataRetriever.ErrBadMaxNumBytesPerSenderUpperBound)
 	}
 
+	if args.TxCacheBoundsConfig.MaxTrackedBlocks == 0 {
```
Comment (andreibancioiu) [inline dataRetriever/txpool/argShardedTxPool.go:58]: Generally speaking, such checks should be done by the component that uses them (e.g. the underlying selection tracker). However, I see, this is our current convention (seeing other checks above), can stay as it is.

Comment (andreibancioiu) [reply dataRetriever/txpool/argShardedTxPool.go:58]: Do a small check, locally, though. Drop this check, and see if the error is still properly detected and reported.

File: txcache/selectionTracker.go [hunk 3]
```diff
@@ -100,6 +104,16 @@ func (st *selectionTracker) OnProposedBlock(
 	}
 
 	st.blocks[string(blockHash)] = tBlock
+
+	if len(st.blocks) == int(st.maxTrackedBlocks) {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply txcache/selectionTracker.go]: Not applicable comment (in my analysis).

File: txcache/selectionTracker.go [hunk 4]
```diff
@@ -100,6 +104,16 @@ func (st *selectionTracker) OnProposedBlock(
...
+		log.Warn("selectionTracker.OnProposedBlock: max tracked blocks reached",
+			"len(st.blocks)", len(st.blocks),
+		)
+
+		// clear all the proposed blocks
+		st.blocks = make(map[string]*trackedBlock)
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply txcache/selectionTracker.go]: Discussed internally. This is what we want, at this moment.

File: integrationTests/chainSimulator/mempool/mempool_test.go [hunk 5]
```diff
@@ -893,6 +896,173 @@ func Test_Selection_ShouldNotSelectSameTransactionsWithManyTransactions(t *testi
...
+			Nonce:    3,
+			PrevHash: []byte("blockHash2"),
+			RootHash: []byte(fmt.Sprintf("rootHash%d", 1)),
+		},
+		selectionSession,
+		defaultBlockchainInfo,
+	)
+	require.Nil(t, err)
+
+	// do the second selection (the rest of the transactions should be selected)
```
Comment (andreibancioiu) [inline integrationTests/chainSimulator/mempool/mempool_test.go:1034]: :+1:

File: integrationTests/chainSimulator/mempool/mempool_test.go [hunk 6]
```diff
@@ -1056,6 +1227,366 @@ func Test_Selection_ShouldNotSelectSameTransactionsWithManyTransactionsAndExecut
...
+		30_000,
+		int(selectionLoopMaximumDuration.Milliseconds()),
+		10,
+	)
+
+	nonceTracker := newNoncesTracker()
+
+	// create txs and add them to txpool
+	numTxs := numTxsPerSender * len(senders)
+	for i := 0; i < numTxsPerSender; i++ {
```
Comment (andreibancioiu) [inline integrationTests/chainSimulator/mempool/mempool_test.go:1280]: Not sure, do we have some utilities in `integrationTests/chainSimulator/mempool/testutils_test.go` that can help with having these new tests a bit shorter (more concise)? If not, that's all right, we can refactor the tests (make them more accessible) in the future.

File: integrationTests/chainSimulator/mempool/mempool_test.go [hunk 7]
```diff
@@ -1056,6 +1227,366 @@ func Test_Selection_ShouldNotSelectSameTransactionsWithManyTransactionsAndExecut
...
+	// no transactions should be returned
+	selectedTransactions, _ = txpool.SelectTransactions(selectionSession, options, blockchainInfo)
+	require.Equal(t, 0, len(selectedTransactions))
+
+	for _, tx := range proposedTxs {
+		require.True(t, txpool.RemoveTxByHash(tx))
+	}
+}
+
+func Test_Selection_MaxTrackedBlocksReached(t *testing.T) {
```
Comment (andreibancioiu) [inline integrationTests/chainSimulator/mempool/mempool_test.go:1430]: :+1:

File: txcache/errors.go (hunk 1 above)
Comment (danielradu10) [reply txcache/errors.go]: fixed

File: integrationTests/chainSimulator/mempool/mempool_test.go (hunk 6 above)
Comment (danielradu10) [reply integrationTests/chainSimulator/mempool/mempool_test.go:1280]: at the moment, for these tests... no... 
only for the benchmarks we have used some utilities... 
I will add a TODO here

File: dataRetriever/txpool/argShardedTxPool.go (hunk 2 above)
Comment (danielradu10) [reply dataRetriever/txpool/argShardedTxPool.go:58]: I think that because the SelectionTracker is only initialized in TxCache and the TxCache in this method - **createTxCache** - the error will not be detected, because in case of an error from NewTxCache, we just return a **disabled cache.** 

```
func (txPool *shardedTxPool) createTxCache(cacheID string) txCache {

...
		cache, err := txcache.NewTxCache(config, txPool.host)
		if err != nil {
			log.Error("shardedTxPool.createTxCache()", "err", err)
			return txcache.NewDisabledCache()
		}
...
}
```

Comment (danielradu10) [reply dataRetriever/txpool/argShardedTxPool.go:58]: added a TODO for now

Comment (andreibancioiu) [reply dataRetriever/txpool/argShardedTxPool.go:58]: You are right :+1:

**Review by andreibancioiu — APPROVED** [review]

File: cmd/node/config/config.toml [hunk 8]
```diff
@@ -423,6 +423,7 @@
 
 [TxCacheBounds]
     MaxNumBytesPerSenderUpperBound = 33_554_432             #32MB
+    MaxTrackedBlocks = 100
```
Comment (raduchis) [inline cmd/node/config/config.toml:426]: 1 minute, this means the difference between the current proposed block and the last executed block would be max 100 blocks?

File: txcache/selectionTracker.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply txcache/selectionTracker.go]: yes, as long as not 2 or more blocks can be added at once it should not be a problem.

File: txcache/selectionTracker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply txcache/selectionTracker.go]: it might be better to either stop the processing or to just keep the first 100 and compute based on those as discussed on the thread

File: dataRetriever/txpool/argShardedTxPool.go (hunk 2 above)
Comment (raduchis) [reply dataRetriever/txpool/argShardedTxPool.go:58]: maybe compare it to another variable -> the allowed max difference between the current proposed block and last execution result. This is currently discussed to be  ~5. A MaxTrackedBlocks < 5 should not be valid. 
Add a TODO pls for the comparison.

File: txcache/selectionTracker.go [hunk 9]
```diff
@@ -10,29 +10,33 @@ import (
...
+	maxTrackedBlocks uint32
 }
 
 // NewSelectionTracker creates a new selectionTracker
-func NewSelectionTracker(txCache txCacheForSelectionTracker) (*selectionTracker, error) {
+func NewSelectionTracker(txCache txCacheForSelectionTracker, maxTrackedBlocks uint32) (*selectionTracker, error) {
 	if check.IfNil(txCache) {
 		return nil, errNilTxCache
 	}
+	if maxTrackedBlocks == 0 {
```
Comment (raduchis) [inline txcache/selectionTracker.go:28]: also here with the current configuration a MaxTrackedBlocks < 5 should not be valid.

File: dataRetriever/txpool/argShardedTxPool.go (hunk 2 above)
Comment (danielradu10) [reply dataRetriever/txpool/argShardedTxPool.go:58]: fixed, added a TODO

File: txcache/selectionTracker.go (hunk 9 above)
Comment (danielradu10) [reply txcache/selectionTracker.go:28]: fixed, added a TODO

File: txcache/selectionTracker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielradu10) [reply txcache/selectionTracker.go]: fixed

File: cmd/node/config/config.toml (hunk 8 above)
Comment (danielradu10) [reply cmd/node/config/config.toml:426]: at the moment, yes

File: dataRetriever/txpool/argShardedTxPool.go [hunk 10]
```diff
@@ -54,6 +54,7 @@ func (args *ArgShardedTxPool) verify() error {
 		return fmt.Errorf("%w: MaxNumBytesPerSenderUpperBound is not valid", dataRetriever.ErrBadMaxNumBytesPerSenderUpperBound)
 	}
 
+	// TODO compare with another variable
```
Comment (andreibancioiu) [inline dataRetriever/txpool/argShardedTxPool.go]: Maybe "compare with the maximum allowed offset between proposing a block and actually executing it". Opinion.

File: testscommon/headerHandlerStub.go [hunk 11]
```diff
@@ -435,3 +439,35 @@ func (hhs *HeaderHandlerStub) SetBlockBodyTypeInt32(blockBodyType int32) error {
...
+
+	return false
+}
+
+func (hhs *HeaderHandlerStub) GetGasLimit() uint32 {
+	if hhs.GetGasLimitCalled != nil {
+		return hhs.GetGasLimitCalled()
+	}
+
+	return 0
```
Comment (andreibancioiu) [inline testscommon/headerHandlerStub.go]: We'll have some conflicts with `rc/supernova`, at some point, but that is fine. We'll handle them when appropriate. By convention, we add those dummy comments on mock methods.

File: txcache/selectionTracker.go [hunk 12]
```diff
@@ -24,6 +24,7 @@ func NewSelectionTracker(txCache txCacheForSelectionTracker, maxTrackedBlocks ui
 	if check.IfNil(txCache) {
 		return nil, errNilTxCache
 	}
+	// TODO compare with another variable
```
Comment (andreibancioiu) [inline txcache/selectionTracker.go]: Same comment as above.

File: txcache/selectionTracker_test.go [hunk 13]
```diff
@@ -353,7 +353,67 @@ func TestSelectionTracker_OnProposedBlockShouldCleanWhenMaxTrackedBlocksIsReache
...
+				{},
+			},
+		},
+		accountsProvider,
+		holders.NewBlockchainInfo([]byte("hash0"), nil, 20),
+	)
+	require.Nil(t, err)
+	require.Equal(t, 4, len(tracker.blocks))
+
+	// this one should be added because is an empty block
```
Comment (andreibancioiu) [inline txcache/selectionTracker_test.go]: :+1:

File: txcache/selectionTracker_test.go [hunk 14]
```diff
@@ -353,7 +353,67 @@ func TestSelectionTracker_OnProposedBlockShouldCleanWhenMaxTrackedBlocksIsReache
...
+			PrevHash: []byte(fmt.Sprintf("hash%d", 3)),
+			RootHash: []byte("rootHash0"),
+		},
+		accountsProvider,
+		holders.NewBlockchainInfo([]byte("hash0"), nil, 20),
+	)
+	require.Equal(t, errBadBlockWhileMaxTrackedBlocksReached, err)
+	require.Equal(t, 3, len(tracker.blocks))
+
+	// this one should be added because it has new execution results
```
Comment (andreibancioiu) [inline txcache/selectionTracker_test.go:380]: :+1:

File: txcache/selectionTracker_test.go [hunk 15]
```diff
@@ -353,7 +353,67 @@ func TestSelectionTracker_OnProposedBlockShouldCleanWhenMaxTrackedBlocksIsReache
 	accountsProvider := txcachemocks.NewAccountNonceAndBalanceProviderMock()
 
 	proposeBlocks(t, numOfBlocks, tracker, accountsProvider)
-	require.Equal(t, 0, len(tracker.blocks))
+
+	// this one should not be added, it doesn't have execution results
```
Comment (andreibancioiu) [inline txcache/selectionTracker_test.go]: "..., but it has miniblocks with transactions" or something similar? Also, for the sake of the test, perhaps add a dummy transaction hash in the dummy miniblock?

File: txcache/selectionTracker.go [hunk 16]
```diff
@@ -140,6 +138,28 @@ func (st *selectionTracker) OnExecutedBlock(blockHeader data.HeaderHandler) erro
 	return nil
 }
 
+// checkReceivedBlockNoLock first checks if MaxTrackedBlocks is reached
+// then, if MaxTrackedBlocks is reached, checks that the received block is as expected
```
Comment (andreibancioiu) [inline txcache/selectionTracker.go]: We can add a more detailed description of the conditions here (opinion).

File: txcache/selectionTracker.go [hunk 17]
```diff
@@ -140,6 +138,28 @@ func (st *selectionTracker) OnExecutedBlock(blockHeader data.HeaderHandler) erro
...
+func (st *selectionTracker) checkReceivedBlockNoLock(blockBody *block.Body, blockHeader data.HeaderHandler) error {
+	if len(st.blocks) < int(st.maxTrackedBlocks) {
+		return nil
+	}
+
+	if len(blockBody.MiniBlocks) != 0 && len(blockHeader.GetExecutionResultsHandlers()) == 0 {
```
Comment (andreibancioiu) [inline txcache/selectionTracker.go]: Maybe extract some easy-to-read variables such as `noNewTransactions` and `noNewExecutionResults`? Just an idea.

File: txcache/selectionTracker.go [hunk 18]
```diff
@@ -140,6 +138,28 @@ func (st *selectionTracker) OnExecutedBlock(blockHeader data.HeaderHandler) erro
...
+		log.Warn("selectionTracker.checkReceivedBlockNoLock: received bad block while max tracked blocks is reached. "+
+			"should receive empty block or a block with new execution results",
+			"len(st.blocks)", len(st.blocks),
+		)
+
+		return errBadBlockWhileMaxTrackedBlocksReached
+	}
+
+	log.Warn("selectionTracker.checkReceivedBlockNoLock: max tracked blocks reached",
```
Comment (andreibancioiu) [inline txcache/selectionTracker.go]: "..., but ..." (something about tolerance to empty blocks / execution-recovery blocks).

File: dataRetriever/txpool/argShardedTxPool.go (hunk 10 above)
Comment (danielradu10) [reply dataRetriever/txpool/argShardedTxPool.go]: fixed

File: dataRetriever/txpool/argShardedTxPool.go [hunk 19]
```diff
@@ -49,9 +49,15 @@ func (args *ArgShardedTxPool) verify() error {
...
 
+	// TODO brainstorm if these checks could be moved in the underlying component
 	if args.TxCacheBoundsConfig.MaxNumBytesPerSenderUpperBound == 0 {
 		return fmt.Errorf("%w: MaxNumBytesPerSenderUpperBound is not valid", dataRetriever.ErrBadMaxNumBytesPerSenderUpperBound)
 	}
 
+	// TODO compare with another variable
+	if args.TxCacheBoundsConfig.MaxTrackedBlocks == 0 {
+		return fmt.Errorf("%w: MaxTrackedBlocks is not valid", dataRetriever.ErrBadMaxTrackedBlocks)
+	}
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply dataRetriever/txpool/argShardedTxPool.go]: fixed

File: testscommon/headerHandlerStub.go (hunk 11 above)
Comment (danielradu10) [reply testscommon/headerHandlerStub.go]: forgot to add the dummy comments, fixed now.

File: txcache/selectionTracker.go [hunk 20]
```diff
@@ -10,29 +10,34 @@ import (
...
+	maxTrackedBlocks uint32
 }
 
 // NewSelectionTracker creates a new selectionTracker
-func NewSelectionTracker(txCache txCacheForSelectionTracker) (*selectionTracker, error) {
+func NewSelectionTracker(txCache txCacheForSelectionTracker, maxTrackedBlocks uint32) (*selectionTracker, error) {
 	if check.IfNil(txCache) {
 		return nil, errNilTxCache
 	}
+	// TODO compare with another variable
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply txcache/selectionTracker.go]: fixed

File: txcache/selectionTracker.go (hunk 12 above)
Comment (danielradu10) [reply txcache/selectionTracker.go]: fixed

File: txcache/selectionTracker.go (hunk 16 above)
Comment (danielradu10) [reply txcache/selectionTracker.go]: fixed

File: txcache/selectionTracker.go [hunk 21]
```diff
@@ -126,6 +138,28 @@ func (st *selectionTracker) OnExecutedBlock(blockHeader data.HeaderHandler) erro
...
 }
 
+// checkReceivedBlockNoLock first checks if MaxTrackedBlocks is reached
+// then, if MaxTrackedBlocks is reached, checks that the received block is as expected
+func (st *selectionTracker) checkReceivedBlockNoLock(blockBody *block.Body, blockHeader data.HeaderHandler) error {
+	if len(st.blocks) < int(st.maxTrackedBlocks) {
+		return nil
+	}
+
+	if len(blockBody.MiniBlocks) != 0 && len(blockHeader.GetExecutionResultsHandlers()) == 0 {
```
[the remark this answers is not in the document]
Comment (danielradu10) [reply txcache/selectionTracker.go]: fixed, added variables

File: txcache/selectionTracker.go (hunk 17 above)
Comment (danielradu10) [reply txcache/selectionTracker.go]: fixed

File: txcache/selectionTracker.go (hunk 18 above)
Comment (danielradu10) [reply txcache/selectionTracker.go]: fixed

File: txcache/selectionTracker_test.go (hunk 15 above)
Comment (danielradu10) [reply txcache/selectionTracker_test.go]: fixed

File: txcache/selectionTracker.go [hunk 22]
```diff
@@ -139,13 +139,16 @@ func (st *selectionTracker) OnExecutedBlock(blockHeader data.HeaderHandler) erro
...
-// then, if MaxTrackedBlocks is reached, checks that the received block is as expected
+// if MaxTrackedBlocks is reached, the received block must either have an empty body or contain new execution results.
 func (st *selectionTracker) checkReceivedBlockNoLock(blockBody *block.Body, blockHeader data.HeaderHandler) error {
 	if len(st.blocks) < int(st.maxTrackedBlocks) {
 		return nil
 	}
 
-	if len(blockBody.MiniBlocks) != 0 && len(blockHeader.GetExecutionResultsHandlers()) == 0 {
+	hasNewTransactions := len(blockBody.MiniBlocks) != 0
+	noNewExecutionResults := len(blockHeader.GetExecutionResultsHandlers()) == 0
```
Comment (andreibancioiu) [inline txcache/selectionTracker.go:149]: `hasNoNewExecutionResults`.

File: dataRetriever/txpool/argShardedTxPool.go (hunk 2 above)
Comment (raduchis) [reply dataRetriever/txpool/argShardedTxPool.go:58]: ok!

File: txcache/selectionTracker.go (hunk 9 above)
Comment (raduchis) [reply txcache/selectionTracker.go:28]: ok
